Support @DataBoundSetter for triggers and publishers

g-dx opened the pull request:
Currently jenkins jobs defined using the generated DSL for the `gitlab-plugin` are required to supply all possible values for any defined triggers or publishers due to the use of the `@DataBoundConstructor` annotation. This leads to the problem described in #613 where any modifications to the trigger or publisher constructor signatures cause existing job definitions to fail when run against a new version of the plugin.

Fix this by adding support for a setter driven approach to configure triggers and publishers with the `@DataBoundSetter` annotation. This allows all unimportant configuration to be ignored when defining jobs and prevents new functionality from breaking existing jobs when upgrading.

christ66 commented on src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java:
```diff
@@ -103,9 +103,10 @@
 
 
     @DataBoundConstructor
-    @GeneratePojoBuilder(intoPackage = "*.builder.generated", withFactoryMethod = "*")
```
I'm not familiar with this annotation, however I believe this is generating code which might break backward compatibility.

g-dx replied:
Good point. Will leave as is.

christ66 commented on src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java:
```diff
@@ -103,9 +103,10 @@
-    						 TriggerOpenMergeRequest triggerOpenMergeRequestOnPush, boolean triggerOnNoteRequest, String noteRegex, 
+    public GitLabPushTrigger() { }
+
+    public GitLabPushTrigger(boolean triggerOnPush, boolean triggerOnMergeRequest, boolean triggerOnAcceptedMergeRequest, boolean triggerOnClosedMergeRequest,
```
I would recommend deprecating the constructors and adding a javadoc to recommend using getter/setter methods.

g-dx replied:
Done.

christ66 commented on src/main/java/com/dabsquared/gitlabjenkins/publisher/GitLabMessagePublisher.java:
```diff
@@ -35,7 +36,6 @@
     private String unstableNoteText;
 
-    @DataBoundConstructor
     public GitLabMessagePublisher(boolean onlyForFailure, boolean replaceSuccessNote, boolean replaceFailureNote, boolean replaceAbortNote, boolean replaceUnstableNote,
```
Probably best to deprecate the constructor.

g-dx replied:
Done.

g-dx commented:
PTAL.

omehegan commented:
@christ66 can you take another look at this one?

christ66 commented on src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java:
```diff
@@ -101,11 +102,11 @@
     private transient MergeRequestLabelFilter mergeRequestLabelFilter;
 
-
+    @Deprecated
```
Did something happen to the empty constructor you created? I was only mentioning to use the `@deprecated` annotation only if there was a new constructor which should be used instead of this one, and this one should be deprecated in favor of the new constructor + using setter methods. If the other empty constructor is not needed, then you should remove the deprecated annotation. I also believe that adding a `@deprecated` should require javadoc suggesting to use the other constructor.

g-dx replied:
Ah, apologies. I stripped the merge back to only what's required to support setters and left the rest for fear of affecting backwards compatibility.

I will add no-arg constructors where necessary and update `@Deprecated` constructors to inform clients to use the no-arg version in conjunction with setters.

christ66 commented on src/main/java/com/dabsquared/gitlabjenkins/publisher/GitLabMessagePublisher.java:
```diff
@@ -35,6 +36,10 @@
+     * @deprecated use {@link #GitLabMessagePublisher()} with setters to configure an instance of this class.
+     */
+    @Deprecated
     @DataBoundConstructor
```
Should the annotation `@DataBoundConstructor` be around the `public GitLabMessagePublisher() { }` constructor instead of the current constructor.

g-dx replied:
I must admit I'm somewhat unsure exactly where `@DataBoundConstructor` is actually required now that `@DataBoundSetter` has been added. 

When making these changes I used [this](https://github.com/jenkinsci/slack-plugin/pull/236/files) merge from the `slack-plugin` as a reference. Looking again at those changes they do appear to have moved `@DataBoundConstructor` annotation onto the no-arg constructors. As such I will make the same change.

g-dx committed:
Support @DataBoundSetter for triggers and publishers.

Currently jenkins jobs defined using the generated DSL for the `gitlab-plugin` are required to supply all possible values for any defined triggers or publishers due to the use of the `@DataBoundConstructor` annotation. This leads to the problem described in https://github.com/jenkinsci/gitlab-plugin/issues/613 where any modifications to the trigger or publisher constructor signatures cause existing job definitions to fail when run against a new version of the plugin.

Fix this by adding support for a setter driven approach to configure triggers and publishers with the `@DataBoundSetter` annotation. This allows all unimportant configuration to be ignored when defining jobs and prevents new functionality from breaking existing jobs when upgrading.

christ66 reviewed (approved):
LGTM 👍

g-dx commented:
@christ66, @omehegan - if these changes are ok any ETA on when these changes will find their way into a release?

omehegan commented:
@g-dx right now I am holding releases until we can get the GitLab v4 API support complete. That is being tracked in #614.

g-dx commented:
No problem. Thanks for the update.

omehegan commented:
This has just been released in plugin version 1.5.2. It will be in the update center within a few hours.

jakub-bochenski commented:
Doesn't seem to work, created #674
